src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/HandlerOrchestrator.cs

```diff
@@ -0,0 +1,95 @@
+        private InternalConnectionContext _internalConnectionContext;
+        private int _routingCount;
+        private List<ConnectionHandlerContext> _routingHistory;
+
+        public HandlerOrchestrator(IHandler<HandlerRequest> initialHandler, ConnectionHandlerContext initialConnectionHandlerContext, InternalConnectionContext internalConnectionContext, List<ConnectionHandlerContext> history)
+        {
+            _initialHandler = initialHandler;
+            _initialConnectionHandlerContext = initialConnectionHandlerContext;
+            _internalConnectionContext = internalConnectionContext;
+            _routingHistory = history;
+            _routingCount = 0;
+        }
+
+        /// <summary>
+        /// Intiates the pre-defined chain of handlers
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        /// <exception cref="HandlerNotFoundException"></exception>
+        public async Task ProcessRequestAsync(bool isAsync, CancellationToken ct) 
+        {
+            CreateHandlerChain();
+            if (_initialHandler != null)
+            {
+                await _initialHandler.Handle(_initialConnectionHandlerContext, isAsync, ct).ConfigureAwait(false);
+            }
+            else
+            {
+                //TODO: Deal with missing handler exception in the connector and populate relevant SqlException in future.
+                throw new HandlerNotFoundException();
+            }
+        }
+
+        /// <summary>
+        /// Creates the chain of handlers in the order definition
+        /// </summary>
+        private void CreateHandlerChain()
+        {
+            // List of handlers to be added to the chain, ex : DSH -> TCH -> Prelogin -> Login -> null
+            List<IHandler<ConnectionHandlerContext>> handlers = new List<IHandler<ConnectionHandlerContext>>
+            {
+                new DataSourceParsingHandler(),
+                new TransportCreationHandler(),
+                //new PreLoginHandler(),
+            };
+
+            for (int i = 0; i < handlers.Count - 1; i++)
+            {
+                handlers[i].NextHandler = handlers[i + 1];
+            }
+
+            // Set the first handler in the chain
+            if (handlers.Count > 0)
+            {
+                _initialHandler = (IHandler<HandlerRequest>)handlers[0];
+            }
+        }
+
+        /// <summary>
+        /// Adds the context of previous handler into the routingHistory object in case
+        /// of a reroute, exception etc.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <exception cref="ArgumentNullException"></exception>
+        private void AddToHistory(ConnectionHandlerContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context), "Context cannot be null");
```

Comment:
Nit: resource string

Reply:
Well this seems more like a Debug assert than an exception. If this exception is every thrown, then the driver imlpementation is wrong